.github/workflows/test.yaml

```diff
@@ -11,89 +10,147 @@ on:
   schedule:
     - cron: '0 20 * * SUN'
 
+concurrency:
+  group: ${{ github.workflow }}-${{ github.event.pull_request.number || github.ref }}
+  cancel-in-progress: true
+
+defaults:
+  run:
+    shell: bash -e {0}
+
+env:
+  DISPLAY: ":99.0"
+  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+  COV: "--cov=./nbsite --cov-report=xml"
+
 jobs:
   pre_commit:
     name: Run pre-commit
     runs-on: "ubuntu-latest"
     steps:
       - uses: holoviz-dev/holoviz_tasks/pre-commit@v0
-  test_suite:
-    name: Test on ${{ matrix.python-version }}, ${{ matrix.os }}
-    needs: [pre_commit]
-    runs-on: ${{ matrix.os }}
-    strategy:
-      fail-fast: false
-      matrix:
-        os: ['ubuntu-latest']
-        python-version: ['3.9', '3.10', '3.11', '3.12']
-    timeout-minutes: 30
-    defaults:
-      run:
-        shell: bash -l {0}
-    env:
-      PYTHON_VERSION: ${{ matrix.python-version }}
-      GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+  setup:
+    name: Setup workflow
+    runs-on: ubuntu-latest
+    permissions:
+      pull-requests: read
+    outputs:
+      code_change: ${{ steps.filter.outputs.code }}
+      matrix: ${{ env.MATRIX }}
     steps:
       - uses: actions/checkout@v4
+        if: github.event_name != 'pull_request'
+      - name: Check for code changes
+        uses: dorny/paths-filter@v3
+        id: filter
         with:
-          fetch-depth: "100"
-      - uses: actions/setup-python@v5
-        with:
-          python-version: ${{ matrix.python-version }}
-      - name: Fetch
-        run: git fetch --prune --tags
-      - name: env setup
+          filters: |
+            code:
+              - 'nbsite/**'
+              - 'pixi.toml'
+              - 'pyproject.toml'
+              - '.github/workflows/test.yaml'
+      - name: Set matrix option
+        run: |
+          if [[ '${{ github.event_name }}' == 'workflow_dispatch' ]]; then
+            OPTION=${{ github.event.inputs.target }}
+          elif [[ '${{ github.event_name }}' == 'schedule' ]]; then
+            OPTION="full"
+          elif [[ '${{ github.event_name }}' == 'push' && '${{ github.ref_type }}' == 'tag' ]]; then
+            OPTION="full"
+          else
+            OPTION="default"
+          fi
+          echo "MATRIX_OPTION=$OPTION" >> $GITHUB_ENV
+      - name: Set test matrix with 'default' option
+        if: env.MATRIX_OPTION == 'default'
+        run: |
+          MATRIX=$(jq -nsc '{
+              "os": ["ubuntu-latest", "macos-latest", "windows-latest"],
+              "environment": ["test-39", "test-312"],
+            }')
+          echo "MATRIX=$MATRIX" >> $GITHUB_ENV
+      - name: Set test matrix with 'full' option
+        if: env.MATRIX_OPTION == 'full'
         run: |
-          set -xe
-          python -VV
-          python -m pip install --upgrade pip
-          python -m pip install -e ".[tests]"
-      - name: pip list
-        run: pip list
-      - name: tests
-        run: pytest nbsite
-  build_docs:
```

Comment:
The docs were built and deployed to the dev site on purpose on pushes to PRs. It's useful to check the visual effects of some changes and also serves as a kind of regression test. Is this re-implemented somewhere else or was this entirely removed?

Reply:
My bad, I actually thought it was redundant code since we can always trigger the docs build manually :(

Reply:
> Do you recommend I implement it back?

Yes definitely, it's very useful.

Jean-Luc told me once about Chesterton's fence, since then I always have that concept in mind.